### PR TITLE
cs: optimize pattern matching on authentic structs

### PR DESCRIPTION
**Pull request by yjqww6**
This eliminates the branch of `unsafe-struct-ref` on cptypes level. 

A benchmark:
https://gist.github.com/yjqww6/d20134c86a749cf19a1946e8a2aaecbc

Results (on a M1 Mac):
master
```
cpu time: 1458 real time: 1481 gc time: 261
420000
```
this
```
cpu time: 1088 real time: 1107 gc time: 333
420000
```

Alternative approach may be making `match` generate `unsafe-struct*-ref` directly.

### CODE REVIEW

**Comment by samth** [comment]
Is authentic-ness available in static structure information? If so, match could generate that pretty easily. Even so, this seems like a good idea.

**Comment by yjqww6** [comment]
> Is authentic-ness available in static structure information? If so, match could generate that pretty easily. Even so, this seems like a good idea.

It doesn't currently. Also, the authentic-ness is not really known in compile time, because it could be determined dynamically:
```
> (let ()
    (define a prop:authentic)
    (struct A (a) #:property a #t)
    (struct-type-authentic? struct:A))
#t
```

**Comment by samth** [comment]
Right, we're not going to have a complete check, but a sound approximation (roughly just using `#: authentic`) ought to be enough to capture the important optimizations.

**Review by mflatt — APPROVED** [review]
Looks good to me (didn't actually try running it, yet), so waiting for a review by @gus-massa to check/confirm.

**Comment by gus-massa** [comment]
It looks good, but I never read the details of the tricks to reduce records in `cp0`, so I'm confused by

    (define (rtd-obviously-incompatible? x y)
      (let ([x-flds (rtd-flds x)]
            [y-flds (rtd-flds y)])
        (or (and (fixnum? x-flds) (not (fixnum? y-flds)))
            (and (not (fixnum? x-flds)) (fixnum? y-flds)))))

I tried to create manually a few `rtd`s but I got `vector`s instead of `fixnum`s.

About the three passes for `cp0`, I found a similar case in https://github.com/cisco/ChezScheme/pull/421

**Comment by yjqww6** [comment]
@gus-massa , that function is a simplified version of
https://github.com/racket/racket/blob/63aa90c7487c361ce4786f79ae2d5fca1984da78/racket/src/ChezScheme/s/cp0.ss#L3960-L3971

`rtd`s with anonymous fields (created by `make-record-type-descriptor*`) would have `fixnum` in `flds`. That's the case of racket structs.
`impersonator`s are defined with `define-record`, and have named fields (list of vectors in `flds`).

**Comment by gus-massa** [comment]
LGTM

**Review by gus-massa — APPROVED** [review]

**Comment by pmatos** [comment]
Note that this change causes the solo chez workflows to fail.

**Comment by yjqww6** [comment]
> Note that this change causes the solo chez workflows to fail.

The failure should be due to outdated expected error message in https://github.com/racket/racket/commit/0523a5311ce95ab9658a585e59c907433db9fdf9 .

**Comment by mflatt** [comment]
I'll fix the tests (that I broke) in a separate push.

**Pull request merged**
